ouroboros-consensus-cardano/src/tools/Cardano/Tools/DBTruncater/Run.hs

```diff
@@ -0,0 +1,109 @@
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Cardano.Tools.DBTruncater.Run (truncate) where
+
+import           Cardano.Slotting.Slot (SlotNo, WithOrigin (..))
+import           Cardano.Tools.DBAnalyser.HasAnalysis
+import           Cardano.Tools.DBTruncater.Types
+import           Control.Monad
+import           Control.Tracer
+import           Data.Functor.Identity
+import           Ouroboros.Consensus.Block.Abstract (HeaderFields (..),
+                     getBlockHeaderFields)
+import           Ouroboros.Consensus.Config
+import           Ouroboros.Consensus.Node as Node
+import           Ouroboros.Consensus.Node.InitStorage as Node
+import           Ouroboros.Consensus.Storage.Common
+import           Ouroboros.Consensus.Storage.ImmutableDB (ImmutableDB, Iterator,
+                     IteratorResult (..))
+import qualified Ouroboros.Consensus.Storage.ImmutableDB as ImmutableDB
+import           Ouroboros.Consensus.Storage.ImmutableDB.Impl
+import           Ouroboros.Consensus.Util.IOLike
+import           Ouroboros.Consensus.Util.ResourceRegistry (runWithTempRegistry,
+                     withRegistry)
+import           Prelude hiding (truncate)
+import           System.IO
+
+truncate :: forall block .
+            ( Node.RunNode block, HasProtocolInfo block )
+         => DBTruncaterConfig -> Args block -> IO ()
+truncate DBTruncaterConfig{ dbDir, truncatePoint, verbose } args = do
+  withRegistry $ \registry -> do
+    lock <- mkLock
+    immutableDBTracer <- mkTracer lock verbose
+    ProtocolInfo {
+      pInfoConfig = config
+    } <- mkProtocolInfo args
+    let
+      TruncatePoint truncateSlotNo = truncatePoint
+      fs = Node.stdMkChainDbHasFS dbDir (RelativeMountPoint "immutable")
+      chunkInfo = Node.nodeImmutableDbChunkInfo (configStorage config)
+      immutableDBArgs :: ImmutableDbArgs Identity IO block
+      immutableDBArgs =
+        (ImmutableDB.defaultArgs fs)
+          { immTracer = immutableDBTracer
+          , immRegistry = registry
+          , immCheckIntegrity = nodeCheckIntegrity (configStorage config)
+          , immCodecConfig = configCodec config
+          , immChunkInfo = chunkInfo
+          }
+
+    withDB immutableDBArgs $ \(immutableDB, internal) -> do
+      iterator <- ImmutableDB.streamAll immutableDB registry
+        ((,,) <$> GetSlot <*> GetBlock <*> GetIsEBB)
```

Comment:
> make it faster by not streaming every single block from the DB

Yeah, I imagine this would be nice; of course, the "dream" would be constant time truncation; but maybe, once you only read from the primary/secondary index (so removing `GetBlock`, it is already fast enough for e.g. the mainnet chain?

Reply:
I've swapped out `GetBlock` for `GetHeader` (I'd missed `getHeaderFields`) -- am I correct in thinking that the headers are accessible entirely from the index?

Reply:
I also anticipate that runtime linear in the length of the chain is fine for the anticipated uses of this tool.

On the other hand, it does seem relatively easy to reach logarithmic time: do a binary search over the block numbers (modulo EBBs :'() to find the greatest block number whose slot is below the requested truncation threshold.

Reply:
> am I correct in thinking that the headers are accessible entirely from the index?

No, the (secondary) index just stores the header offset+size as well as the header hash, see [Section 8.2.2 in the report](https://input-output-hk.github.io/ouroboros-network/pdfs/report/report.pdf#subsection.8.2.2). But using `GetHeader` instead of `GetBlock` should already result in way less disk IO as you don't have to read the full block (on mainnet, we have 88KiB max block size and 1.1KB header size).

> On the other hand, it does seem relatively easy to reach logarithmic time

Conceptually, constant time is simple, as the ImmDB uses a uniform slot chunk size (see https://github.com/input-output-hk/ouroboros-network/issues/4512), so finding the respective chunk file is just a division; the only obstacle is whether this is very annoying to implement given the existing API/how much one would have to expose in addition.